# CM 3010

## 1 Load Datasets

In [35]:
import numpy as np
import os

#### load the all dataset and  check the first file called " health expenditure"

In [36]:
import pandas as pd
# Specify the file path 
file_path_health = 'Health_Expenditure/health.csv'
file_path_working_age = 'Age_Dependency_Working_Age/working.csv'
file_path_youth = 'Age_Dependency_Youth/youth.csv'
file_path_elderly = 'Age_Dependency_Elderly/elderly.csv'

# Read the CSV file, skipping the first 4 rows
df = pd.read_csv(file_path_health, skiprows=4)

# Display the first few rows of the DataFrame to verify the structure
print("First few rows of the health expenditure Dataframe:")
display(df.head())


First few rows of the health expenditure Dataframe:


Country Name Country Code  \
0                        Aruba          ABW   
1  Africa Eastern and Southern          AFE   
2                  Afghanistan          AFG   
3   Africa Western and Central          AFW   
4                       Angola          AGO   

                          Indicator Name     Indicator Code  1960  1961  1962  \
0  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   NaN   NaN   NaN   
1  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   NaN   NaN   NaN   
2  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   NaN   NaN   NaN   
3  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   NaN   NaN   NaN   
4  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   NaN   NaN   NaN   

   1963  1964  1965  ...       2015       2016       2017       2018  \
0   NaN   NaN   NaN  ...        NaN        NaN        NaN        NaN   
1   NaN   NaN   NaN  ...   5.998256   6.098205   5.954076   5.809503   
2   NaN   NaN   NaN  ...  10.105348  11.818590  12.620817  14.208419   
3   NaN   NaN   NaN  ...   3.805882   3.796499   3.816799   3.430094   
4   NaN   NaN   NaN  ...   2.605795   2.713150   2.793838   2.588796   

        2019       2020       2021  2022  2023  Unnamed: 68  
0        NaN        NaN        NaN   NaN   NaN          NaN  
1   5.830176   5.881698   5.865601   NaN   NaN          NaN  
2  14.831320  15.533614  21.827950   NaN   NaN          NaN  
3   3.417091   3.838112   4.141484   NaN   NaN          NaN  
4   2.475180   3.220706   2.960106   NaN   NaN          NaN  

[5 rows x 69 columns]

In [37]:
# Check the column names to ensure they are correctly parsed
print("\nColumn names:")
print(df.columns)


Column names:
Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022',
       '2023', 'Unnamed: 68'],
      dtype='object')


In [38]:
# Accessing specific columns or years' data
# Adjust column names as needed based on the actual DataFrame structure
years = ['2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022', '2023']
for year in years:
    if year in df.columns:
        year_data = df[year]
        print(f"\nValue counts for '{year}' column:")
        print(year_data.value_counts())
    else:
        print(f"\n'{year}' column not found in DataFrame.")



Value counts for '2013' column:
2013
5.097014    2
3.567699    2
4.965594    2
8.870000    1
6.125854    1
           ..
8.437964    1
5.626529    1
5.771889    1
3.802313    1
7.109403    1
Name: count, Length: 231, dtype: int64

Value counts for '2014' column:
2014
7.050000    2
3.470834    2
5.840839    1
8.386758    1
5.380247    1
           ..
7.891412    1
4.773536    1
5.888568    1
3.947554    1
8.132273    1
Name: count, Length: 232, dtype: int64

Value counts for '2015' column:
2015
3.452434    2
4.973738    2
5.998256    1
5.470000    1
3.673638    1
           ..
8.220045    1
4.646756    1
6.006020    1
4.022648    1
7.449827    1
Name: count, Length: 232, dtype: int64

Value counts for '2016' column:
2016
6.367661     2
3.384545     2
5.041881     2
10.060000    1
4.580086     1
            ..
2.859417     1
8.448747     1
4.439100     1
6.070082     1
7.466447     1
Name: count, Length: 230, dtype: int64

Value counts for '2017' column:
2017
6.351296     2
2.963604    

#### check the second dataset called "working age "

In [39]:
# Read the CSV file, skipping the first 4 rows
df_age_dependency = pd.read_csv(file_path_working_age, skiprows=0)

# Display the column names to verify the structure
print("Columns in the Age Dependency DataFrame:")
print(df_age_dependency.columns)

# Display the first few rows of the DataFrame to verify the structure
print("\nFirst few rows of the Age Dependency DataFrame:")
display(df_age_dependency.head())

# Create a mapping dictionary for renaming columns
rename_mapping = {col: col.split(' ')[0] for col in df_age_dependency.columns if 'YR' in col}

# Apply the renaming
df_age_dependency.rename(columns=rename_mapping, inplace=True)

# Verify if specific columns exist
required_columns = ['Country Name', 'Country Code'] + [str(year) for year in range(2013, 2024)]
missing_columns = [col for col in required_columns if col not in df_age_dependency.columns]

if missing_columns:
    print(f"Missing columns: {missing_columns}")
else:
    print("All required columns are present.")

# Display the updated column names to verify the renaming
print("\nUpdated columns in the Age Dependency DataFrame:")
print(df_age_dependency.columns)

# Display the first few rows of the updated DataFrame to verify the structure
print("\nFirst few rows of the updated Age Dependency DataFrame:")
display(df_age_dependency.head())



Columns in the Age Dependency DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code',
       '1990 [YR1990]', '2000 [YR2000]', '2014 [YR2014]', '2015 [YR2015]',
       '2016 [YR2016]', '2017 [YR2017]', '2018 [YR2018]', '2019 [YR2019]',
       '2020 [YR2020]', '2021 [YR2021]', '2022 [YR2022]', '2023 [YR2023]'],
      dtype='object')

First few rows of the Age Dependency DataFrame:


Series Name  Series Code  \
0  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
1  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
2  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
3  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
4  Age dependency ratio (% of working-age populat...  SP.POP.DPND   

     Country Name Country Code  1990 [YR1990]  2000 [YR2000]  2014 [YR2014]  \
0     Afghanistan          AFG      98.890488     108.346635      94.706697   
1         Albania          ALB      62.078455      59.657295      46.434431   
2         Algeria          DZA      85.054610      61.500880      50.679694   
3  American Samoa          ASM      75.260226      76.473619      58.221980   
4         Andorra          AND      37.500963      38.879667      38.970817   

   2015 [YR2015]  2016 [YR2016]  2017 [YR2017]  2018 [YR2018]  2019 [YR2019]  \
0      93.042448      92.012125      90.543094      89.091451      87.649322   
1      46.413119      46.552893      46.723001      46.914934      47.280397   
2      51.961774      53.315437      54.679122      55.953864      57.053988   
3      57.348527      56.495843      55.627222      54.727290      53.839402   
4      39.066892      39.088835      38.658429      38.130927      37.958328   

   2020 [YR2020]  2021 [YR2021]  2022 [YR2022] 2023 [YR2023]  
0      85.959550      84.599257      83.570360            ..  
1      47.696632      48.183223      48.813511            ..  
2      57.889467      58.463731      58.808385            ..  
3      53.028094      52.335690      52.463928            ..  
4      37.987924      38.140600      38.431923            ..

Missing columns: ['2013']

Updated columns in the Age Dependency DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code', '1990',
       '2000', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021',
       '2022', '2023'],
      dtype='object')

First few rows of the updated Age Dependency DataFrame:


Series Name  Series Code  \
0  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
1  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
2  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
3  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
4  Age dependency ratio (% of working-age populat...  SP.POP.DPND   

     Country Name Country Code       1990        2000       2014       2015  \
0     Afghanistan          AFG  98.890488  108.346635  94.706697  93.042448   
1         Albania          ALB  62.078455   59.657295  46.434431  46.413119   
2         Algeria          DZA  85.054610   61.500880  50.679694  51.961774   
3  American Samoa          ASM  75.260226   76.473619  58.221980  57.348527   
4         Andorra          AND  37.500963   38.879667  38.970817  39.066892   

        2016       2017       2018       2019       2020       2021  \
0  92.012125  90.543094  89.091451  87.649322  85.959550  84.599257   
1  46.552893  46.723001  46.914934  47.280397  47.696632  48.183223   
2  53.315437  54.679122  55.953864  57.053988  57.889467  58.463731   
3  56.495843  55.627222  54.727290  53.839402  53.028094  52.335690   
4  39.088835  38.658429  38.130927  37.958328  37.987924  38.140600   

        2022 2023  
0  83.570360   ..  
1  48.813511   ..  
2  58.808385   ..  
3  52.463928   ..  
4  38.431923   ..

#### check the third data called age dependency youth

In [40]:
# Specify the file path
file_path_youth = 'Age_Dependency_Youth/youth.csv'

# Read the CSV file
df_age_dependency_youth = pd.read_csv(file_path_youth)

# Display the column names to verify the structure
print("Columns in the Youth Age Dependency DataFrame:")
print(df_age_dependency_youth.columns)

# Display the first few rows of the DataFrame to verify the structure
print("\nFirst few rows of the Youth Age Dependency DataFrame:")
display(df_age_dependency_youth.head())

# Create a mapping dictionary for renaming columns
rename_mapping = {col: col.split(' ')[0] for col in df_age_dependency_youth.columns if 'YR' in col}

# Apply the renaming
df_age_dependency_youth.rename(columns=rename_mapping, inplace=True)

# Verify if specific columns exist
required_columns = ['Country Name', 'Country Code'] + [str(year) for year in range(2013, 2024)]
missing_columns = [col for col in required_columns if col not in df_age_dependency_youth.columns]

if missing_columns:
    print(f"Missing columns: {missing_columns}")
else:
    print("All required columns are present.")

# Display the updated column names to verify the renaming
print("\nUpdated columns in the Youth Age Dependency DataFrame:")
print(df_age_dependency_youth.columns)

# Display the first few rows of the updated DataFrame to verify the structure
print("\nFirst few rows of the updated Youth Age Dependency DataFrame:")
display(df_age_dependency_youth.head())


Columns in the Youth Age Dependency DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code',
       '1990 [YR1990]', '2000 [YR2000]', '2014 [YR2014]', '2015 [YR2015]',
       '2016 [YR2016]', '2017 [YR2017]', '2018 [YR2018]', '2019 [YR2019]',
       '2020 [YR2020]', '2021 [YR2021]', '2022 [YR2022]', '2023 [YR2023]'],
      dtype='object')

First few rows of the Youth Age Dependency DataFrame:


Series Name     Series Code  \
0  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
1  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
2  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
3  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
4  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   

     Country Name Country Code     1990 [YR1990]     2000 [YR2000]  \
0     Afghanistan          AFG  94.4222572629881   103.58632038782   
1         Albania          ALB  52.7784284725856  47.1689375465814   
2         Algeria          DZA  79.2949742930899  54.6473546144873   
3  American Samoa          ASM  69.6891951326785  70.8609873624875   
4         Andorra          AND  24.4590466901101  21.2174059210388   

      2014 [YR2014]     2015 [YR2015]     2016 [YR2016]     2017 [YR2017]  \
0  90.0159003100492  88.3982020260321  87.4057740333038  85.9704074964568   
1  27.2542318244552  26.5983978791543  26.0774898968204  25.5931941158779   
2  43.0501410944391  44.0738078725369   45.128959468372  46.1725227632222   
3  50.7529618519528  49.5619677755315  48.3465690532324  47.0880659472045   
4  22.1515057435579  21.5599620088775  21.1032922363047  20.3770821205236   

      2018 [YR2018]     2019 [YR2019]     2020 [YR2020]     2021 [YR2021]  \
0  84.5398440923121  83.1124205721626  81.4649968705518  80.1613851925191   
1  25.0975887781106   24.672046977961  24.3278873157337  24.1267907154218   
2  47.1106452784713  47.8584188063625  48.3885154333489  48.6675037389368   
3  45.7743553695242  44.4473342002601  43.1520010601643  41.9104962283936   
4  19.4696620937299  18.8607983808302  18.3999289646599  18.0608963154583   

      2022 [YR2022] 2023 [YR2023]  
0  79.1750593586432            ..  
1  24.0283580946095            ..  
2  48.6640452550879            ..  
3  41.3891662936052            ..  
4  17.7115307909752            ..

Missing columns: ['2013']

Updated columns in the Youth Age Dependency DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code', '1990',
       '2000', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021',
       '2022', '2023'],
      dtype='object')

First few rows of the updated Youth Age Dependency DataFrame:


Series Name     Series Code  \
0  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
1  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
2  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
3  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
4  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   

     Country Name Country Code              1990              2000  \
0     Afghanistan          AFG  94.4222572629881   103.58632038782   
1         Albania          ALB  52.7784284725856  47.1689375465814   
2         Algeria          DZA  79.2949742930899  54.6473546144873   
3  American Samoa          ASM  69.6891951326785  70.8609873624875   
4         Andorra          AND  24.4590466901101  21.2174059210388   

               2014              2015              2016              2017  \
0  90.0159003100492  88.3982020260321  87.4057740333038  85.9704074964568   
1  27.2542318244552  26.5983978791543  26.0774898968204  25.5931941158779   
2  43.0501410944391  44.0738078725369   45.128959468372  46.1725227632222   
3  50.7529618519528  49.5619677755315  48.3465690532324  47.0880659472045   
4  22.1515057435579  21.5599620088775  21.1032922363047  20.3770821205236   

               2018              2019              2020              2021  \
0  84.5398440923121  83.1124205721626  81.4649968705518  80.1613851925191   
1  25.0975887781106   24.672046977961  24.3278873157337  24.1267907154218   
2  47.1106452784713  47.8584188063625  48.3885154333489  48.6675037389368   
3  45.7743553695242  44.4473342002601  43.1520010601643  41.9104962283936   
4  19.4696620937299  18.8607983808302  18.3999289646599  18.0608963154583   

               2022 2023  
0  79.1750593586432   ..  
1  24.0283580946095   ..  
2  48.6640452550879   ..  
3  41.3891662936052   ..  
4  17.7115307909752   ..

In [41]:
#### check the forth data called "age dependency eladerly"

In [42]:
# Read the CSV file, skipping the first 4 rows
df_age_dependency_elderly = pd.read_csv(file_path_elderly)

# Display the column names to verify the structure
print("Columns in the e Dependency DataFrame:")
print(df_age_dependency_elderly.columns)

# Display the first few rows of the DataFrame to verify the structure
print("\nFirst few rows of the Old Age Dependency DataFrame:")
display(df_age_dependency_elderly.head())

# Create a mapping dictionary for renaming columns
rename_mapping = {col: col.split(' ')[0] for col in df_age_dependency_elderly.columns if 'YR' in col}

# Apply the renaming
df_age_dependency_elderly.rename(columns=rename_mapping, inplace=True)

# Verify if specific columns exist
required_columns = ['Country Name', 'Country Code'] + [str(year) for year in range(2013, 2024)]
missing_columns = [col for col in required_columns if col not in df_age_dependency_elderly.columns]

if missing_columns:
    print(f"Missing columns: {missing_columns}")
else:
    print("All required columns are present.")

# Display the updated column names to verify the renaming
print("\nUpdated columns in the old Age Dependency DataFrame:")
print(df_age_dependency_elderly.columns)

# Display the first few rows of the updated DataFrame to verify the structure
print("\nFirst few rows of the updated Old Age Dependency DataFrame:")
display(df_age_dependency_elderly.head())

Columns in the e Dependency DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code',
       '1990 [YR1990]', '2000 [YR2000]', '2013 [YR2013]', '2014 [YR2014]',
       '2015 [YR2015]', '2016 [YR2016]', '2017 [YR2017]', '2018 [YR2018]',
       '2019 [YR2019]', '2020 [YR2020]', '2021 [YR2021]', '2022 [YR2022]'],
      dtype='object')

First few rows of the Old Age Dependency DataFrame:


Series Name     Series Code                    Country Name  \
0  Age dependency ratio, old  SP.POP.DPND.OL     Africa Eastern and Southern   
1  Age dependency ratio, old  SP.POP.DPND.OL      Africa Western and Central   
2  Age dependency ratio, old  SP.POP.DPND.OL                      Arab World   
3  Age dependency ratio, old  SP.POP.DPND.OL          Caribbean small states   
4  Age dependency ratio, old  SP.POP.DPND.OL  Central Europe and the Baltics   

  Country Code     1990 [YR1990]     2000 [YR2000]     2013 [YR2013]  \
0          AFE  5.63253104272149  5.46284227621415  5.47361175756582   
1          AFW  6.24127149621416  5.85597363363416  5.57374386723284   
2          ARB  6.21773874825515  6.27995575239571  6.03990626550959   
3          CSS  10.1788035385631  9.11479302845442  9.79748203212278   
4          CEB  17.0822103596391  19.7217506820497  23.4299256962288   

      2014 [YR2014]     2015 [YR2015]     2016 [YR2016]     2017 [YR2017]  \
0  5.50991761850398  5.54414747774553  5.58459980363562  5.63177710618492   
1  5.51992116730695  5.47653435510672  5.45287144428005  5.43269882188918   
2  6.15106787123626  6.26831619212666  6.39060642863966  6.52419869518265   
3  9.96872324015086  10.1736138483477  10.4366555980686  10.7146915370276   
4  24.2014483288317  25.0178731410971  25.8742735761673  26.7566239577768   

      2018 [YR2018]     2019 [YR2019]     2020 [YR2020]     2021 [YR2021]  \
0  5.67701712388147  5.71842422318594  5.73311054063037  5.70619942558857   
1  5.41843841270947  5.41072397825298  5.38630996805807  5.33700290542415   
2  6.67103562856093  6.82660882514592  6.99140851692453  7.14443368453384   
3  10.9685165119656  11.2916544510961  11.6870137653663  12.0172234145172   
4  27.6532057102598  28.5838599501776  29.4678238952773  30.1503278273799   

      2022 [YR2022]  
0  5.66235268493612  
1  5.29982718151313  
2  7.27801082080313  
3  12.3607072835352  
4  29.7803934152905

Missing columns: ['2023']

Updated columns in the old Age Dependency DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code', '1990',
       '2000', '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020',
       '2021', '2022'],
      dtype='object')

First few rows of the updated Old Age Dependency DataFrame:


Series Name     Series Code                    Country Name  \
0  Age dependency ratio, old  SP.POP.DPND.OL     Africa Eastern and Southern   
1  Age dependency ratio, old  SP.POP.DPND.OL      Africa Western and Central   
2  Age dependency ratio, old  SP.POP.DPND.OL                      Arab World   
3  Age dependency ratio, old  SP.POP.DPND.OL          Caribbean small states   
4  Age dependency ratio, old  SP.POP.DPND.OL  Central Europe and the Baltics   

  Country Code              1990              2000              2013  \
0          AFE  5.63253104272149  5.46284227621415  5.47361175756582   
1          AFW  6.24127149621416  5.85597363363416  5.57374386723284   
2          ARB  6.21773874825515  6.27995575239571  6.03990626550959   
3          CSS  10.1788035385631  9.11479302845442  9.79748203212278   
4          CEB  17.0822103596391  19.7217506820497  23.4299256962288   

               2014              2015              2016              2017  \
0  5.50991761850398  5.54414747774553  5.58459980363562  5.63177710618492   
1  5.51992116730695  5.47653435510672  5.45287144428005  5.43269882188918   
2  6.15106787123626  6.26831619212666  6.39060642863966  6.52419869518265   
3  9.96872324015086  10.1736138483477  10.4366555980686  10.7146915370276   
4  24.2014483288317  25.0178731410971  25.8742735761673  26.7566239577768   

               2018              2019              2020              2021  \
0  5.67701712388147  5.71842422318594  5.73311054063037  5.70619942558857   
1  5.41843841270947  5.41072397825298  5.38630996805807  5.33700290542415   
2  6.67103562856093  6.82660882514592  6.99140851692453  7.14443368453384   
3  10.9685165119656  11.2916544510961  11.6870137653663  12.0172234145172   
4  27.6532057102598  28.5838599501776  29.4678238952773  30.1503278273799   

               2022  
0  5.66235268493612  
1  5.29982718151313  
2  7.27801082080313  
3  12.3607072835352  
4  29.7803934152905

## 2. Processing the Raw data

In [49]:

# Load the CSV file
file_path = 'World_population/population.csv'
population_df = pd.read_csv(file_path)

population_df = population_df.dropna(subset=['Code'])
population_df = population_df[population_df['Code'].str.strip() != '']

# Filter the DataFrame to keep only the years from 2014 to 2021
filtered_population_df = population_df[(population_df['Year'] >= 2014) & (population_df['Year'] <= 2021)]

# Save the filtered DataFrame to a new CSV file
filtered_csv_path = 'World_population/filtered_population_2014_2021.csv'
filtered_population_df.to_csv(filtered_csv_path, index=False)

print(f"Filtered data saved to {filtered_csv_path}")


Filtered data saved to World_population/filtered_population_2014_2021.csv


### 2.1 only selected years are required and drop the unnecessary years

In [43]:
# Drop columns for years 1960 to 2012
years_to_drop = [str(year) for year in range(1960, 2014)]
df.drop(columns=years_to_drop, inplace=True, errors='ignore')
df.drop(columns='Unnamed: 68', inplace=True, errors='ignore')
df_age_dependency.drop(columns=years_to_drop, inplace=True, errors='ignore')
df_age_dependency_youth.drop(columns=years_to_drop, inplace=True, errors='ignore')
df_age_dependency_elderly.drop(columns=years_to_drop, inplace=True, errors='ignore')

#drop 2023 seperatly
df.drop(columns='2023', inplace=True, errors='ignore')
df_age_dependency.drop(columns='2023', inplace=True, errors='ignore')
df_age_dependency_youth.drop(columns='2023', inplace=True, errors='ignore')
df_age_dependency_elderly.drop(columns='2023', inplace=True, errors='ignore')

# Display the first few rows of the DataFrame to verify the structure
print("First few rows of the DataFrame after cleaning:")
display(df.head())
display(df_age_dependency.head())

# Check the remaining column names
print("\nRemaining column names:")
print(df.columns)


First few rows of the DataFrame after cleaning:


Country Name Country Code  \
0                        Aruba          ABW   
1  Africa Eastern and Southern          AFE   
2                  Afghanistan          AFG   
3   Africa Western and Central          AFW   
4                       Angola          AGO   

                          Indicator Name     Indicator Code      2014  \
0  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS       NaN   
1  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  5.840839   
2  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  9.528878   
3  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  3.583262   
4  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  2.434129   

        2015       2016       2017       2018       2019       2020  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1   5.998256   6.098205   5.954076   5.809503   5.830176   5.881698   
2  10.105348  11.818590  12.620817  14.208419  14.831320  15.533614   
3   3.805882   3.796499   3.816799   3.430094   3.417091   3.838112   
4   2.605795   2.713150   2.793838   2.588796   2.475180   3.220706   

        2021  2022  
0        NaN   NaN  
1   5.865601   NaN  
2  21.827950   NaN  
3   4.141484   NaN  
4   2.960106   NaN

Series Name  Series Code  \
0  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
1  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
2  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
3  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
4  Age dependency ratio (% of working-age populat...  SP.POP.DPND   

     Country Name Country Code       2014       2015       2016       2017  \
0     Afghanistan          AFG  94.706697  93.042448  92.012125  90.543094   
1         Albania          ALB  46.434431  46.413119  46.552893  46.723001   
2         Algeria          DZA  50.679694  51.961774  53.315437  54.679122   
3  American Samoa          ASM  58.221980  57.348527  56.495843  55.627222   
4         Andorra          AND  38.970817  39.066892  39.088835  38.658429   

        2018       2019       2020       2021       2022  
0  89.091451  87.649322  85.959550  84.599257  83.570360  
1  46.914934  47.280397  47.696632  48.183223  48.813511  
2  55.953864  57.053988  57.889467  58.463731  58.808385  
3  54.727290  53.839402  53.028094  52.335690  52.463928  
4  38.130927  37.958328  37.987924  38.140600  38.431923


Remaining column names:
Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022'],
      dtype='object')


In [44]:
display(df_age_dependency_youth.head())
display(df_age_dependency_elderly.head())


Series Name     Series Code  \
0  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
1  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
2  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
3  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
4  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   

     Country Name Country Code              2014              2015  \
0     Afghanistan          AFG  90.0159003100492  88.3982020260321   
1         Albania          ALB  27.2542318244552  26.5983978791543   
2         Algeria          DZA  43.0501410944391  44.0738078725369   
3  American Samoa          ASM  50.7529618519528  49.5619677755315   
4         Andorra          AND  22.1515057435579  21.5599620088775   

               2016              2017              2018              2019  \
0  87.4057740333038  85.9704074964568  84.5398440923121  83.1124205721626   
1  26.0774898968204  25.5931941158779  25.0975887781106   24.672046977961   
2   45.128959468372  46.1725227632222  47.1106452784713  47.8584188063625   
3  48.3465690532324  47.0880659472045  45.7743553695242  44.4473342002601   
4  21.1032922363047  20.3770821205236  19.4696620937299  18.8607983808302   

               2020              2021              2022  
0  81.4649968705518  80.1613851925191  79.1750593586432  
1  24.3278873157337  24.1267907154218  24.0283580946095  
2  48.3885154333489  48.6675037389368  48.6640452550879  
3  43.1520010601643  41.9104962283936  41.3891662936052  
4  18.3999289646599  18.0608963154583  17.7115307909752

Series Name     Series Code                    Country Name  \
0  Age dependency ratio, old  SP.POP.DPND.OL     Africa Eastern and Southern   
1  Age dependency ratio, old  SP.POP.DPND.OL      Africa Western and Central   
2  Age dependency ratio, old  SP.POP.DPND.OL                      Arab World   
3  Age dependency ratio, old  SP.POP.DPND.OL          Caribbean small states   
4  Age dependency ratio, old  SP.POP.DPND.OL  Central Europe and the Baltics   

  Country Code              2014              2015              2016  \
0          AFE  5.50991761850398  5.54414747774553  5.58459980363562   
1          AFW  5.51992116730695  5.47653435510672  5.45287144428005   
2          ARB  6.15106787123626  6.26831619212666  6.39060642863966   
3          CSS  9.96872324015086  10.1736138483477  10.4366555980686   
4          CEB  24.2014483288317  25.0178731410971  25.8742735761673   

               2017              2018              2019              2020  \
0  5.63177710618492  5.67701712388147  5.71842422318594  5.73311054063037   
1  5.43269882188918  5.41843841270947  5.41072397825298  5.38630996805807   
2  6.52419869518265  6.67103562856093  6.82660882514592  6.99140851692453   
3  10.7146915370276  10.9685165119656  11.2916544510961  11.6870137653663   
4  26.7566239577768  27.6532057102598  28.5838599501776  29.4678238952773   

               2021              2022  
0  5.70619942558857  5.66235268493612  
1  5.33700290542415  5.29982718151313  
2  7.14443368453384  7.27801082080313  
3  12.0172234145172  12.3607072835352  
4  30.1503278273799  29.7803934152905

From the data shown above, we can see that for certain data only 2014 data onwards available. At the same time, the data for health expenditure in 2022 and onwards is missing. Hence we only take all the database from 2014 to 2021.

### check for missing value

In [31]:
# Clean and Check for missing values
df = df.replace({np.nan: 0})
missing_values = df.isnull().sum()
print("\nMissing values in each column:")
print(missing_values)


Missing values in each column:
Country Name      0
Country Code      0
Indicator Name    0
Indicator Code    0
2014              0
2015              0
2016              0
2017              0
2018              0
2019              0
2020              0
2021              0
2022              0
dtype: int64


In [47]:
# Function to process the dataframes
def process_dataframe(data, skiprows):
    # Read the CSV file
    df = data
    
    df = df.replace({np.nan: 0})
    
    # Check for missing values
    missing_values = df.isnull().sum()
    print(missing_values)
    
    # Drop rows with any missing values
    df_cleaned = df.dropna()
    
    # Display the first few rows of the DataFrame to verify the structure
    #print(f"\nFirst few rows of the DataFrame after dropping rows with missing values for {file_path}:")
    display(df_cleaned.head())
    
    return df_cleaned

# Process each dataframe
df_health_cleaned = process_dataframe(df, skiprows=0)
df_working_age_cleaned = process_dataframe(df_age_dependency, skiprows=0)
df_youth_cleaned = process_dataframe(df_age_dependency_youth, skiprows=0)
df_elderly_cleaned = process_dataframe(df_age_dependency_elderly, skiprows=0)

# Check the remaining column names for each DataFrame
print("\nRemaining column names in the health expenditure DataFrame:")
print(df_health_cleaned.columns)

print("\nRemaining column names in the age dependency working age DataFrame:")
print(df_working_age_cleaned.columns)

print("\nRemaining column names in the age dependency youth DataFrame:")
print(df_youth_cleaned.columns)

print("\nRemaining column names in the age dependency elderly DataFrame:")
print(df_elderly_cleaned.columns)

Country Name      0
Country Code      0
Indicator Name    0
Indicator Code    0
2014              0
2015              0
2016              0
2017              0
2018              0
2019              0
2020              0
2021              0
2022              0
dtype: int64


Country Name Country Code  \
0                        Aruba          ABW   
1  Africa Eastern and Southern          AFE   
2                  Afghanistan          AFG   
3   Africa Western and Central          AFW   
4                       Angola          AGO   

                          Indicator Name     Indicator Code      2014  \
0  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  0.000000   
1  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  5.840839   
2  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  9.528878   
3  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  3.583262   
4  Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  2.434129   

        2015       2016       2017       2018       2019       2020  \
0   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   
1   5.998256   6.098205   5.954076   5.809503   5.830176   5.881698   
2  10.105348  11.818590  12.620817  14.208419  14.831320  15.533614   
3   3.805882   3.796499   3.816799   3.430094   3.417091   3.838112   
4   2.605795   2.713150   2.793838   2.588796   2.475180   3.220706   

        2021  2022  
0   0.000000   0.0  
1   5.865601   0.0  
2  21.827950   0.0  
3   4.141484   0.0  
4   2.960106   0.0

Series Name     0
Series Code     0
Country Name    0
Country Code    0
2014            0
2015            0
2016            0
2017            0
2018            0
2019            0
2020            0
2021            0
2022            0
dtype: int64


Series Name  Series Code  \
0  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
1  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
2  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
3  Age dependency ratio (% of working-age populat...  SP.POP.DPND   
4  Age dependency ratio (% of working-age populat...  SP.POP.DPND   

     Country Name Country Code       2014       2015       2016       2017  \
0     Afghanistan          AFG  94.706697  93.042448  92.012125  90.543094   
1         Albania          ALB  46.434431  46.413119  46.552893  46.723001   
2         Algeria          DZA  50.679694  51.961774  53.315437  54.679122   
3  American Samoa          ASM  58.221980  57.348527  56.495843  55.627222   
4         Andorra          AND  38.970817  39.066892  39.088835  38.658429   

        2018       2019       2020       2021       2022  
0  89.091451  87.649322  85.959550  84.599257  83.570360  
1  46.914934  47.280397  47.696632  48.183223  48.813511  
2  55.953864  57.053988  57.889467  58.463731  58.808385  
3  54.727290  53.839402  53.028094  52.335690  52.463928  
4  38.130927  37.958328  37.987924  38.140600  38.431923

Series Name     0
Series Code     0
Country Name    0
Country Code    0
2014            0
2015            0
2016            0
2017            0
2018            0
2019            0
2020            0
2021            0
2022            0
dtype: int64


Series Name     Series Code  \
0  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
1  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
2  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
3  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   
4  Age dependency ratio, young (% of working-age ...  SP.POP.DPND.YG   

     Country Name Country Code              2014              2015  \
0     Afghanistan          AFG  90.0159003100492  88.3982020260321   
1         Albania          ALB  27.2542318244552  26.5983978791543   
2         Algeria          DZA  43.0501410944391  44.0738078725369   
3  American Samoa          ASM  50.7529618519528  49.5619677755315   
4         Andorra          AND  22.1515057435579  21.5599620088775   

               2016              2017              2018              2019  \
0  87.4057740333038  85.9704074964568  84.5398440923121  83.1124205721626   
1  26.0774898968204  25.5931941158779  25.0975887781106   24.672046977961   
2   45.128959468372  46.1725227632222  47.1106452784713  47.8584188063625   
3  48.3465690532324  47.0880659472045  45.7743553695242  44.4473342002601   
4  21.1032922363047  20.3770821205236  19.4696620937299  18.8607983808302   

               2020              2021              2022  
0  81.4649968705518  80.1613851925191  79.1750593586432  
1  24.3278873157337  24.1267907154218  24.0283580946095  
2  48.3885154333489  48.6675037389368  48.6640452550879  
3  43.1520010601643  41.9104962283936  41.3891662936052  
4  18.3999289646599  18.0608963154583  17.7115307909752

Series Name     0
Series Code     0
Country Name    0
Country Code    0
2014            0
2015            0
2016            0
2017            0
2018            0
2019            0
2020            0
2021            0
2022            0
dtype: int64


Series Name     Series Code                    Country Name  \
0  Age dependency ratio, old  SP.POP.DPND.OL     Africa Eastern and Southern   
1  Age dependency ratio, old  SP.POP.DPND.OL      Africa Western and Central   
2  Age dependency ratio, old  SP.POP.DPND.OL                      Arab World   
3  Age dependency ratio, old  SP.POP.DPND.OL          Caribbean small states   
4  Age dependency ratio, old  SP.POP.DPND.OL  Central Europe and the Baltics   

  Country Code              2014              2015              2016  \
0          AFE  5.50991761850398  5.54414747774553  5.58459980363562   
1          AFW  5.51992116730695  5.47653435510672  5.45287144428005   
2          ARB  6.15106787123626  6.26831619212666  6.39060642863966   
3          CSS  9.96872324015086  10.1736138483477  10.4366555980686   
4          CEB  24.2014483288317  25.0178731410971  25.8742735761673   

               2017              2018              2019              2020  \
0  5.63177710618492  5.67701712388147  5.71842422318594  5.73311054063037   
1  5.43269882188918  5.41843841270947  5.41072397825298  5.38630996805807   
2  6.52419869518265  6.67103562856093  6.82660882514592  6.99140851692453   
3  10.7146915370276  10.9685165119656  11.2916544510961  11.6870137653663   
4  26.7566239577768  27.6532057102598  28.5838599501776  29.4678238952773   

               2021              2022  
0  5.70619942558857  5.66235268493612  
1  5.33700290542415  5.29982718151313  
2  7.14443368453384  7.27801082080313  
3  12.0172234145172  12.3607072835352  
4  30.1503278273799  29.7803934152905


Remaining column names in the health expenditure DataFrame:
Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022'],
      dtype='object')

Remaining column names in the age dependency working age DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code', '2014',
       '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022'],
      dtype='object')

Remaining column names in the age dependency youth DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code', '2014',
       '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022'],
      dtype='object')

Remaining column names in the age dependency elderly DataFrame:
Index(['Series Name', 'Series Code', 'Country Name', 'Country Code', '2014',
       '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022'],
      dtype='object')


## 3. setup MySQL environment

In [14]:
# check MySQL service status
!brew services list

Name  Status  User         File
mysql started tlanchaokeai ~/Library/LaunchAgents/homebrew.mxcl.mysql.plist


In [15]:
pip install sqlalchemy==2.0.20

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.0 -> 24.1
[notice] To update, run: /Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [16]:
# install sql magic, required for running sql commands in jupyter notebook

In [17]:
pip install ipython-sql

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.0 -> 24.1
[notice] To update, run: /Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


# enable sql magic in the notebook
%reload_ext sql
# queries are returned in the pandas format
%config SqlMagic.autopandas=True


pip install mysql-connector-python ipython-sql


In [18]:
%load_ext sql
database_url = "sqlite:///HealthAgeComparisonDB.db"
%sql $database_url


In [19]:
%%sql
CREATE TABLE IF NOT EXISTS HealthExpenditure (
    country TEXT,
    year INTEGER,
    expenditure_percent_gdp REAL
);

CREATE TABLE IF NOT EXISTS AgeDependency (
    country TEXT,
    year INTEGER,
    dependency_ratio REAL
);


 * sqlite:///HealthAgeComparisonDB.db
Done.
Done.


[]

# project name
os.environ["PROJECT_NAME"] = "HealthAgeData"
# database name
os.environ["DATABASE_NAME"] = os.getenv("Health_expenditure")


# mysql credentials
os.environ["MYSQL_USER"] = "SG"
os.environ["MYSQL_USER_PASSWORD"] = "SG_health"

# Import required modules
import os
import sqlite3
from sqlalchemy import create_engine

# Set environment variables
os.environ["PROJECT_NAME"] = "HealthAgeComparison"
os.environ["DATABASE_NAME"] = "HealthAgeComparisonDB.db"

# Load the SQL extension
%load_ext sql

# Connect to the SQLite database
database_url = f"sqlite:///{os.getenv('DATABASE_NAME')}"
%sql $database_url

# Create tables and insert data
%%sql
CREATE TABLE IF NOT EXISTS HealthExpenditure (
    country TEXT,
    year INTEGER,
    expenditure_percent_gdp REAL
);

CREATE TABLE IF NOT EXISTS AgeDependency (
    country TEXT,
    year INTEGER,
    dependency_ratio REAL
);

-- Example of inserting data into HealthExpenditure
INSERT INTO HealthExpenditure (country, year, expenditure_percent_gdp) VALUES
('CountryA', 2020, 7.5),
('CountryB', 2020, 6.8);

-- Example of inserting data into AgeDependency
INSERT INTO AgeDependency (country, year, dependency_ratio) VALUES
('CountryA', 2020, 55.4),
('CountryB', 2020, 47.2);

# Query the tables to see the data
%%sql
SELECT * FROM HealthExpenditure;

SELECT * FROM AgeDependency;
